### PR TITLE
fix(chain redirects): Follow up: Changing network in navbar always redirects to Landing page

### LINKED ISSUE #6678: Follow up: Changing network in navbar always redirects to Landing page
### What happened?

Part of: #6416
Followup for #6613

Pages that still lack the enhancement: 
Profile page
Transfer page
On-Chain Identity

### PR DESCRIPTION
**Pull request by hassnian**
**Thank you for your contribution** to the [KodaDot - One Stop Shop for Polkadot NFTs](https://kodadot.xyz).

👇 __ Let's make a quick check before the contribution.

## PR Type

- [x] Bugfix
- [ ] Feature
- [ ] Refactoring

## Context

- [x] Closes #6678
- [x] Fixes: wallet address not updating after chain change

#### Did your issue had any of the "$" label on it?

- [x] Fill up your DOT address: [Payout](https://kodadot.xyz/dot/transfer?target=16faLfsywwNATaEfbH2ah75dn6ZmctQWpMS5G4KFhbmj5hnD)

#### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

Profile Page
CleanShot 2023-08-13 at 14 50 47

Transfer Page
CleanShot 2023-08-13 at 14 56 14

Identity Page
CleanShot 2023-08-13 at 18 48 17

Drops Page
CleanShot 2023-08-15 at 06 45 21

Teleport Page
CleanShot 2023-08-15 at 06 47 19

Incoming Offers
CleanShot 2023-08-15 at 06 48 02

### <samp>🤖 Generated by Copilot at 11de2ef</samp>

Added chain-specific features for identity and transfer pages. `useChainRedirect` composable and `permision.config.ts` file handle the logic and configuration of showing or hiding these features and redirecting users to the appropriate chain.

### <samp>🤖 Generated by Copilot at 11de2ef</samp>

> _To support different chains with ease_
> _They added some configs and keys_
> _With `useChainRedirect`_
> _They can now detect_
> _The page type and availability_

### CODE REVIEW

**Comment by hassnian** [comment]
got one last small thing that is not working yet , when clicking on the identity button on the sidebar/modal it wont close it , will fix that asap

**Comment by prury** [comment]
Hey @hassnian, 
Things are super low here for me atm(canary not loading, balances not loading), so, i don't know if that's why this is occurring: 
(identity field persists even if i don't have identity on that chain and deposit value also stays unchanged after chain change) 

https://github.com/kodadot/nft-gallery/assets/36627808/99627bbc-7189-42e8-9207-78cd55c908c7

transfer page working nicely ✔️ 
profile page working nicely ✔️

**Comment by hassnian** [comment]
@prury

1.  pls try the identity page again , everything should work , let me know if you find any issues 
2. can we create a new issue for the sidebar bug , when clicking on the identity button , the sidebar doesn't close, I know It's a small one but is not related to this pr

thanks

**Comment by EthVlad** [comment]
Tested on https://deploy-preview-6688--koda-canary.netlify.app/
- Transfer - OK 
- Onchain identity - OK
- Teleport - ❌
- Profile - OK
- Explore - OK
- Drops - ❌

**Comment by hassnian** [comment]
> Tested on https://deploy-preview-6688--koda-canary.netlify.app/
> 
> * Transfer - OK
> * Onchain identity - OK
> * Teleport - ❌
> * Profile - OK
> * Explore - OK
> * Drops - ❌

on it

**Comment by prury** [comment]
@EthVlad a question, let's say i'm on the Drops page on the Kusama Asset hub and i change chains, where it should lead me to? 
i could not stay in the same page because i've just changed chain and it would have to point the to the drop page in the changed chain. does it make sense?

**Comment by prury** [comment]
> 
>     1. pls try the identity page again , everything should work , let me know if you find any issues
> 

Done, it works! ty, sent you a message on telegram about the sidebar bug

**Comment by hassnian** [comment]
@prury pls try again

File: composables/useChainRedirect.ts [hunk 1]
```diff
@@ -1,35 +1,56 @@
...
+  BLOG_SLUG = '/blog/{slug}',
+  PREFIX_MASSMINT = '/{prefix}/massmint',
+  PREFIX_MASSMINT_ONBOARDING = '/{prefix}/massmint/onboarding',
+  PREFIX_CLASSIC_CREATE = '/{prefix}/create',
+  PREFIX_TRANSFER = '/{prefix}/transfer',
+  PREFIX_TELEPORT = '/{prefix}/teleport',
+  PREFIX_ASSETS = '/{prefix}/assets',
+  PREFIX_INCOMING_OFFERS = '/{prefix}/incomingoffers',
+  IDENTITY = '/identity',
+  PROFILE = '/{prefix}/u/{wallet}',
```
Comment (Jarsen136) [inline composables/useChainRedirect.ts:52]: How about 
```suggestion
  PROFILE = '/{prefix}/u/{address}',
```

File: utils/config/permision.config.ts [hunk 2]
```diff
@@ -72,6 +72,90 @@ const hasMassmintCreate: Config<boolean> = {
...
+  rmrk: true,
+  ksm: true,
+  bsx: true,
+  ahk: true,
+  movr: true,
+  glmr: true,
+  snek: true,
+  dot: true,
+  ahp: true,
+}
```
Comment (Jarsen136) [inline utils/config/permision.config.ts:121]:
```suggestion
```

File: utils/config/permision.config.ts [hunk 3]
```diff
@@ -72,6 +72,90 @@ const hasMassmintCreate: Config<boolean> = {
...
+  ksm: false,
+  bsx: true,
+  ahk: false,
+  movr: false,
+  glmr: false,
+  snek: true,
+  dot: false,
+  ahp: false,
+}
+
```
Comment (Jarsen136) [inline utils/config/permision.config.ts:158]: (prefix) => prefix === 'bsx' || prefix === 'snek'

File: utils/config/permision.config.ts [hunk 4]
```diff
@@ -72,6 +72,90 @@ const hasMassmintCreate: Config<boolean> = {
...
+  rmrk: false,
+  ksm: false,
+  bsx: true,
+  ahk: false,
+  movr: false,
+  glmr: false,
+  snek: true,
+  dot: false,
+  ahp: false,
+}
```
Comment (Jarsen136) [inline utils/config/permision.config.ts:145]: (prefix) => prefix === 'bsx' || prefix === 'snek'

File: composables/useChainRedirect.ts [hunk 5]
```diff
@@ -59,118 +94,218 @@ function getEnumKeyByValue<
...
-  [PageType.BLOG]: () => true,
-  [PageType.BLOG_SLUG]: () => true,
+  [RoutePattern.PREFIX_MASSMINT_ONBOARDING]: (chain: Prefix) =>
+    massmintCreateVisible(chain),
+  [RoutePattern.SALES]: (chain: Prefix) => salesVisible(chain),
+  [RoutePattern.HOT]: (chain: Prefix) => hotVisible(chain),
+  [RoutePattern.BLOG]: () => true,
+  [RoutePattern.BLOG_SLUG]: () => true,
+  [RoutePattern.PREFIX_TRANSFER]: (chain: Prefix) => transferVisible(chain),
+  [RoutePattern.PREFIX_TELEPORT]: (chain: Prefix) => teleportVisible(chain),
```
Comment (Jarsen136) [inline composables/useChainRedirect.ts:121]:
```suggestion
  [RoutePattern.PREFIX_TELEPORT]: () => true,
```

**Review by Jarsen136 — COMMENTED** [review]
Please avoid too much unnecessary config.

**Comment by preschian** [comment]
For me, this is too complex. I prefer this way:

Handle only if the pages are under `pages/_prefix`, else, redirect it to home (`/${newPrefix}`). So the pages component not under `pages/_prefix` by default will be redirected to home. 

Let's say we want to support `/identity`. I prefer to move that component from `pages/identity.vue` to `pages/_prefix/identity.vue`

``` js
if (route.params.prefix) {
  // if (route.name === "prefix-u-id") handle address cross chain here
  // if ("prefix-gallery-id" || "prefix-collection-id") redirect to home
  // else replace just the prefix
  router.push({
    params: {
      prefix: newChain,
    },
    query: route.query,
  })
} else {
  router.push({
    path: `/${newChain}`,
  })
}
```

If some of the pages under `pages/_prefix` is still not supported in some chain, I prefer to put the `onBeforeMount()` handle in the pages component. For example, I put this code in `pages/_prefix/drops/index.vue`

```js
onBeforeMount(() => {
  // list not supported prefix handle it here
  if (urlPrefix.value === 'bsx' || urlPrefix.value === 'ksm') {
    router.push('/')
  }
})
```

Or, if the pages are expected only specified in some chain, I prefer to put that outside `/pages/_prefix`. For example, `pages/bsx/offers.vue`

File: composables/useChainRedirect.ts
Comment (preschian) [inline composables/useChainRedirect.ts:1]: https://github.com/kodadot/nft-gallery/pull/6688#issuecomment-1680261066

**Comment by EthVlad** [comment]
@prury Oh I see that drops page is only on Kusama hub. I thought it might be related to specific chain or all unified. In that case its ok that it leads to LP.

File: pages/sales.vue [hunk 6]
```diff
@@ -8,13 +8,25 @@
...
+
 export default {
   name: 'Sales',
   setup() {
     const { urlPrefix } = usePrefix()
+    const router = useRouter()
+
+    const checkRouteAvailability = () => {
+      if (!salesVisible(urlPrefix.value)) {
+        router.push('/')
```
Comment (roiLeo) [inline pages/sales.vue:21]: `navigateTo('/')`

File: pages/_prefix/explore/collectibles.vue [hunk 7]
```diff
@@ -5,9 +5,26 @@
...
   name: 'ExploreCollectibles',
   layout: 'explore-layout',
+
+  setup() {
+    const { urlPrefix } = usePrefix()
+    const router = useRouter()
+
+    const checkRouteAvailability = () => {
+      if (!explorerVisible(urlPrefix.value)) {
+        router.push('/')
```
Comment (roiLeo) [inline pages/_prefix/explore/collectibles.vue:20]:
```suggestion
        navigateTo('/')
```

File: components/drops/Drops.vue [hunk 8]
```diff
@@ -47,10 +47,25 @@ import DropCard from '@/components/drops/DropCard.vue'
...
 
 const { $i18n } = useNuxtApp()
 const drops = useDrops(collectionId)
 const statemintDrops = useDrops(STT_COLLECTION_ID, 'ahp')
+const { urlPrefix } = usePrefix()
+const router = useRouter()
+
+const checkRouteAvailability = () => {
+  if (!dropsVisible(urlPrefix.value)) {
+    router.push('/')
```
Comment (roiLeo) [inline components/drops/Drops.vue:60]:
```suggestion
    navigateTo('/')
```

File: components/drops/Drops.vue [hunk 9]
```diff
@@ -47,10 +47,25 @@ import DropCard from '@/components/drops/DropCard.vue'
 import { collectionId } from '@/components/collection/unlockable/const'
 import { STT_COLLECTION_ID } from '@/components/collection/drop/const'
 import { useDrops } from './useDrops'
+import { dropsVisible } from '~~/utils/config/permision.config'
```
Comment (roiLeo) [inline components/drops/Drops.vue]:
```suggestion
import { dropsVisible } from '@/utils/config/permision.config'
```

File: pages/_prefix/explore/items.vue [hunk 10]
```diff
@@ -18,10 +19,22 @@ export default {
...
     const preferencesStore = usePreferencesStore()
+    const { urlPrefix } = usePrefix()
+    const router = useRouter()
     const isSidebarOpen = computed(
       () => preferencesStore.getsidebarFilterCollapse
     )
 
+    const checkRouteAvailability = () => {
+      if (!explorerVisible(urlPrefix.value)) {
+        router.push('/')
```
Comment (roiLeo) [inline pages/_prefix/explore/items.vue:30]:
```suggestion
        navigateTo('/')
```

**Comment by hassnian** [comment]
@prury had to change some code, feel free to test if everything works :D

File: pages/sales.vue [hunk 11]
```diff
@@ -31,5 +43,8 @@ export default {
       meta: [...this.$seoMeta(metaData)],
     }
   },
+  beforeMount() {
```
Comment (roiLeo) [inline pages/sales.vue:46]: Why it's not in `setup` here?

File: pages/series-insight.vue [hunk 12]
```diff
@@ -25,13 +25,27 @@
...
   components: {
     SeriesTable: () => import('@/components/series/SeriesTable.vue'),
   },
   setup() {
     const { urlPrefix } = usePrefix()
+    const router = useRouter()
+
+    const checkRouteAvailability = () => {
+      if (!seriesInsightVisible(urlPrefix.value)) {
+        router.push('/')
```
Comment (roiLeo) [inline pages/series-insight.vue:41]:
```suggestion
        navigateTo('/')
```

**Review by roiLeo — COMMENTED** [review]
I would use `navigateTo('/')` instead

File: components/navbar/NavbarChainOptions.vue
Comment (preschian) [inline components/navbar/NavbarChainOptions.vue:1]: unused `urlPrefix` on line 19

https://github.com/hassnian/nft-gallery/blob/75d57bd519aba07068a0684c34e1c25a136ef934/components/navbar/NavbarChainOptions.vue#L19

**Review by preschian — APPROVED** [review]

File: components/navbar/NavbarChainOptions.vue
Comment (hassnian) [reply components/navbar/NavbarChainOptions.vue:1]: will fix

**Comment by yangwao** [comment]
Follow up about this one, then? 

https://codeclimate.com/github/kodadot/nft-gallery/pull/6688

image

**Review by roiLeo — APPROVED** [review]

**Comment by yangwao** [comment]
pay 50 usd
builds
image

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $50 @ 4.76 USD/DOT ~ 10.504 $DOT
🧗 [16faLfsywwNATaEfbH2ah75dn6ZmctQWpMS5G4KFhbmj5hnD](https://polkadot.subscan.io/account/16faLfsywwNATaEfbH2ah75dn6ZmctQWpMS5G4KFhbmj5hnD)
🔗 [0x3067b20fe709e088a09d27a07b251c3b049c8f737f55fd58616319fa54298406](https://polkadot.subscan.io/extrinsic/0x3067b20fe709e088a09d27a07b251c3b049c8f737f55fd58616319fa54298406)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Comment by hassnian** [comment]
@yangwao I think something went wrong with the payout 👀
CleanShot 2023-08-17 at 14 45 48@2x

**Comment by vikiival** [comment]
pay 50 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $50 @ 4.77 USD/DOT ~ 10.482 $DOT
🧗 [16faLfsywwNATaEfbH2ah75dn6ZmctQWpMS5G4KFhbmj5hnD](https://polkadot.subscan.io/account/16faLfsywwNATaEfbH2ah75dn6ZmctQWpMS5G4KFhbmj5hnD)
🔗 [0x61e6e80c3cd9270ca425cf9ad6ecf889fe9c3c2b9bcb37d0223734070269599a](https://polkadot.subscan.io/extrinsic/0x61e6e80c3cd9270ca425cf9ad6ecf889fe9c3c2b9bcb37d0223734070269599a)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Comment by yangwao** [comment]
> @yangwao I think something went wrong with the payout 👀

yeah we are caching balances too much and I was topping up balance on bot, yet wasn't that much on address, so it failed and than failed in cascade here :) 

would be nice to check balances on sidebar and check for that like at least once per 30 seconds?

**Pull request merged**